File: android/src/main/kotlin/com/adyen/checkout/flutter/components/blik/BlikComponentFactory.kt
package com.adyen.checkout.flutter.components.blik

import android.content.Context
import androidx.fragment.app.FragmentActivity
import com.adyen.checkout.flutter.components.ComponentPlatformEventHandler
import com.adyen.checkout.flutter.generated.ComponentFlutterInterface
import com.adyen.checkout.flutter.session.CheckoutHolder
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

internal class BlikComponentFactory(
    private val activity: FragmentActivity,
    private val componentFlutterApi: ComponentFlutterInterface,
    private val componentEventHandler: ComponentPlatformEventHandler,
    private val viewTypeId: String,
    private val onDispose: (String) -> Unit,
    private val setCurrentBlikComponent: (BaseBlikComponent) -> Unit,
    private val checkoutHolder: CheckoutHolder? = null,
) : PlatformViewFactory(ComponentFlutterInterface.codec) {
    companion object {
        const val BLIK_COMPONENT_ADVANCED = "blikComponentAdvanced"
        const val BLIK_COMPONENT_SESSION = "blikComponentSession"
    }

    override fun create(
        context: Context,
        viewId: Int,
        args: Any?
    ): PlatformView {
        val creationParams = args as Map<*, *>? ?: emptyMap<Any, Any>()
        val blikComponent =
            if (viewTypeId == BLIK_COMPONENT_SESSION && checkoutHolder != null) {
                BlikSessionComponent(
                    creationParams,
                    activity,
                    componentFlutterApi,
                    componentEventHandler,
                    onDispose,
                    setCurrentBlikComponent,
                    checkoutHolder,
                )
            } else {
                BlikAdvancedComponent(
                    creationParams,
                    activity,
                    componentFlutterApi,
                    componentEventHandler,
                    onDispose,
                    setCurrentBlikComponent,
                )
            }
        setCurrentBlikComponent(blikComponent)
        return blikComponent
    }
}

File: ios/adyen_checkout/Sources/adyen_checkout/components/blik/BlikComponentFactory.swift
import Flutter

class BlikComponentFactory: ComponentFactory {
    static let blikComponentAdvancedId = "blikComponentAdvanced"
    static let blikComponentSessionId = "blikComponentSession"
    let viewTypeId: String
    let checkoutHolder: CheckoutHolder?

    init(
        messenger: FlutterBinaryMessenger,
        componentFlutterApi: ComponentFlutterInterface,
        componentPlatformApi: ComponentPlatformApi,
        viewTypeId: String,
        checkoutHolder: CheckoutHolder? = nil
    ) {
        self.viewTypeId = viewTypeId
        self.checkoutHolder = checkoutHolder
        super.init(
            messenger: messenger,
            componentFlutterApi: componentFlutterApi,
            componentPlatformApi: componentPlatformApi
        )
    }

    override func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        if viewTypeId == BlikComponentFactory.blikComponentSessionId, let checkoutHolder {
            return BlikSessionComponent(
                frame: frame,
                viewIdentifier: viewId,
                arguments: args as? NSDictionary ?? [:],
                binaryMessenger: messenger,
                componentFlutterApi: componentFlutterApi,
                componentPlatformApi: componentPlatformApi,
                checkoutHolder: checkoutHolder
            )
        } else {
            return BlikAdvancedComponent(
                frame: frame,
                viewIdentifier: viewId,
                arguments: args as? NSDictionary ?? [:],
                binaryMessenger: messenger,
                componentFlutterApi: componentFlutterApi,
                componentPlatformApi: componentPlatformApi
            )
        }
    }
}
